File: src/main/java/com/github/ramonwirsch/fopRenderer/RenderConfigExtension.java
package com.github.ramonwirsch.fopRenderer;

import com.github.ramonwirsch.fopRenderer.util.CollectionUtil;
import org.gradle.api.file.FileCollection;

import java.io.File;
import java.util.Map;

/**
 * Created by Ramon on 03.08.2016.
 */
public class RenderConfigExtension {
	private final String name;

	private File rootSrc;

	private File stylesheet;

	private FileCollection dependencies;

	private File resourcesBaseDir;

	private Map<String, Object> resourceCollectionParams = CollectionUtil.immutableMap("exclude", "**/*.xml");

	public RenderConfigExtension(String name) {
		this.name = name;
	}

	public String getName() {
		return name;
	}

	/**
	 * Root source file to render
	 *
	 * @return xml file
	 */
	public File getRootSrc() {
		return rootSrc;
	}

	/**
	 * Root source file to render
	 *
	 * @param src xml file
	 */
	public void setRootSrc(File src) {
		this.rootSrc = src;
	}

	/**
	 * directory, relative to which to find resources (graphics etc...)
	 * <p>
	 * Defaults to parent of {@link #getRootSrc()}
	 *
	 * @return directory
	 */
	public File getResourcesBaseDir() {
		return resourcesBaseDir != null ? resourcesBaseDir : rootSrc.getParentFile();
	}

	/**
	 * directory, relative to which to find resources (graphics etc...)
	 * <p>
	 * Defaults to parent of {@link #getRootSrc()}
	 *
	 * @param resourcesBaseDir directory
	 */
	public void setResourcesBaseDir(File resourcesBaseDir) {
		this.resourcesBaseDir = resourcesBaseDir;
	}

	/**
	 * Stylesheet to use
	 *
	 * @return xsd stylesheet
	 */
	public File getStylesheet() {
		return stylesheet;
	}

	/**
	 * Stylesheet to use
	 *
	 * @param stylesheet xsd stylesheet
	 */
	public void setStylesheet(File stylesheet) {
		this.stylesheet = stylesheet;
	}

	/**
	 * List of files xml files that are included by {@link #getRootSrc()} and whose changes require a new rendering
	 * <p>
	 * Defaults to all files in parent dir of {@link #getRootSrc()}
	 *
	 * @return collection of files
	 */
	public FileCollection getDependencies() {
		return dependencies;
	}

	/**
	 * List of files xml files that are included by {@link #getRootSrc()} and whose changes require a new rendering
	 * <p>
	 * Defaults to all files in parent dir of {@link #getRootSrc()}
	 *
	 * @param dependencies collection of files
	 */
	public void setDependencies(FileCollection dependencies) {
		this.dependencies = dependencies;
	}

	public Map<String, Object> getResourceCollectionParams() {
		return resourceCollectionParams;
	}

	public void setResourceCollectionParams(Map<String, Object> resourceCollectionParams) {
		this.resourceCollectionParams = resourceCollectionParams;
	}
}

File: src/main/kotlin/com/github/ramonwirsch/fopRenderer/RenderConfigExtension.kt
package com.github.ramonwirsch.fopRenderer

import org.gradle.api.Project
import org.gradle.api.file.FileCollection
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.provider.Property
import org.gradle.api.tasks.TaskProvider
import org.gradle.kotlin.dsl.getValue
import org.gradle.kotlin.dsl.property
import org.gradle.kotlin.dsl.setValue
import java.io.File

class RenderConfigExtension(val name: String, project: Project) {

	val rootSrcProperty: RegularFileProperty = project.layout.fileProperty()

	/**
	 * Root source file to render
	 */
	var rootSrc: File
		get() = rootSrcProperty.asFile.get()
		set(value) = rootSrcProperty.set(value)

	/**
	 * Stylesheet to use
	 */
	lateinit var stylesheet: File

	val resourcesBaseDirProperty: Property<File> = project.objects.property<File>().apply { set(rootSrcProperty.asFile.map { it.parentFile }) }

	/**
	 * List of files xml files that are included by [rootSrc] and whose changes require a new rendering
	 *
	 * Defaults to all files in parent dir of [rootSrc]
	 */
	var dependencies: FileCollection = project.fileTree(resourcesBaseDirProperty)

	/**
	 * directory, relative to which to find resources (graphics etc...)
	 *
	 * Defaults to parent of [.getRootSrc]
	 */
	var resourcesBaseDir: File by resourcesBaseDirProperty

	/**
	 * Additional params fed to gradles [org.gradle.api.Project.fileTree] method
	 */
	var resourceCollectionParams: Map<String, Any> = mapOf("exclude" to "**/*.xml")

	/**
	 * whether all validations must have passed before trying to render
	 */
	var isRequiresValidation: Boolean = true

	lateinit var transformTask: TaskProvider<XSLTTransformTask>

	lateinit var renderTask: TaskProvider<FopRenderTask>
}